### PR TITLE
Fix false positives for trailing combinator in selector-combinator-space-after

### LINKED ISSUE #4584: Fix false positives for trailing combinator in selector-combinator-space-after
> Clearly describe the bug

This is valid SCSS:

```scss
.foo ~,
.bar {
  .baz {
    display: none;
  }
}
```

It will compile to the following CSS:

```css
.foo ~ .baz,
.bar .baz {
  display: none;
}
```

However, stylelint is complaining about the first selector, because the _general sibling_ combinator, `~`, is not followed by a single whitespace.

> Which rule, if any, is the bug related to?

`selector-combinator-space-after`

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "selector-combinator-space-after": "always"
  }
}
```

> Which version of stylelint are you using?

11.1.1

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

```
stylelint **/*.scss
```

via `@wordpress/scripts`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No.

> What did you expect to happen?

No error to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

stylelint is complaining about the first selector, because the `~` combinator is not followed by a single whitespace:

```
Expected single space after "~"   selector-combinator-space-after
```

If I were to add a whitespace (which would look strange anyway), stylelint, of course, complains about a stray whitespace before the comma:

```scss
.foo ~ ,
.bar {
  .baz {
    display: none;
  }
}
```

```
Unexpected whitespace before ","   selector-list-comma-space-before
```

This assumes `"selector-list-comma-space-before": "never"` to be included in the rules.

----

This issue is not specific to the `~` combinator, but all of them. When linting SCSS, stylelint should allow for any combinator to be directly followed by a comma, for example, `~,`.

### PR DESCRIPTION
**Pull request by malsf21**
Hey there!

This is a PR that tries to fix the false positive brought up in #4584, namely that a trailing combinator (such as the general sibling combinator, `~`) is not properly handled by `selector-combinator-space-after`.

This PR implements the suggested fix, by adding a check to `isStandardSyntaxSelector` for trailing combinators. It then uses `isStandardSyntaxSelector` in `selectorCombinatorSpaceChecker`, a util used by both `selector-combinator-space-after` and `selector-combinator-space-before`. Finally, it adds the test suggested in the original issue.

> Which issue, if any, is this issue related to?

Closes #4584.

> Is there anything in the PR that needs further explanation?

I wrote a (what I think is poor) implementation of the trailing combinator check:

```js
for (let combinator of punctuationSets.nonSpaceCombinators) {
	if (selector.includes(`${combinator},`)) {
		return false;
	}
}
```

This runs in `O(n)` in the number of `nonSpaceCombinators`, so there aren't any performance benefits from using a set (and I can't imagine running `.includes()` `n` times is that speedy either).

However, I wasn't able to think of another way to do this. Is there something I'm missing? I may have not understood the "search for a trailing combinator" portion of the issue.

### CODE REVIEW

File: lib/rules/selectorCombinatorSpaceChecker.js [hunk 1]
```diff
@@ -17,6 +18,11 @@ module.exports = function (opts) {
...
 		const selector = rule.raws.selector ? rule.raws.selector.raw : rule.selector;
 
+		// Ignore non-standard selectors
+		if (!isStandardSyntaxSelector(selector)) {
+			return;
+		}
+
 		const fixedSelector = parseSelector(selector, opts.result, rule, (selectorTree) => {
 			selectorTree.walkCombinators((node) => {
 				// Ignore spaced descendant combinator
```
Comment (jeddy3) [inline lib/rules/selectorCombinatorSpaceChecker.js:28]: Let's add a `!isStandardSyntaxCombinator(node)` conditional here instead.

**Comment by malsf21** [comment]
Gotcha, I think that makes sense. A bit swamped with work this week but I'll see if I can make the revision!

**Comment by jeddy3** [comment]
> Just to be clear, I should then revert my change to isStandardSyntaxSelector?

That's correct.

> A bit swamped with work this week but I'll see if I can make the revision!

No worries. Only do it if you've time. I'm likely to be swamped with work too going forward, but I'll try to make time for reviewing your pull requests as I enjoy it.

**Comment by malsf21** [comment]
> No worries. Only do it if you've time.

Yup, will do! Aiming for a target of this weekend!

**Comment by malsf21** [comment]
> I think we, therefore, need a new `isStandardSyntaxCombinator` util that will accept a combinator node.

Just to clarify @jeddy3, there's already a util called `isStandardSyntaxCombinator`, which does something else. Should I modify this one, create a new one, do some refactoring, etc?

**Comment by jeddy3** [comment]
> Just to clarify @jeddy3, there's already a util called `isStandardSyntaxCombinator`, which does something else. Should I modify this one, create a new one, do some refactoring, etc? Confused on what you mean by "new util".

My bad, I didn't notice that the util already existed. We'll want to modify this one to include the new conditionals rather than create a new utility.

**Comment by malsf21** [comment]
Implemented the new utility, though I do have a quick question. If I implement the check to just returning false on the first and last nodes, I get an error for this type of test case, involving the SCSS/SASS use case of the nested `>`

```scss
'a { 
  >/*comment*/a {} 
}
```

In this case, the `>` combinator is the first combinator in its container, so the new implementation of the rule would say that it's non-standard, and skip the block. But, the test case in this scenario doesn't like that, and wants a space in between `>` and `a` (ignoring the comment).

In what I've pushed to the branch, I chose to ignore cases of `>`, but I'm not sure if this is the right way to proceed. In this case, should `isStandardSyntaxCombinator` return `true` or `false` for this nested `>`?

```js
if (
  node.parent !== undefined &&
  node.parent !== null &&
  node.value !== '>'
) {
  let parent = node.parent;
  if (node === parent.first) {
    return false;
  }
  if (node === parent.last) {
    return false;
  }
}
```

**Comment by m-allanson** [comment]
@malsf21 it looks like something has gone wonky with the PR, it's showing 131 changed files. Most of them are changes that have already been merged. 

I'm guessing it's caused by the way GitHub doesn't track changes on the target branch after the PR is opened. There are some suggested fixes here: https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch

**Comment by malsf21** [comment]
> @malsf21 it looks like something has gone wonky with the PR, it's showing 131 changed files. Most of them are changes that have already been merged.
> 
> I'm guessing it's caused by the way GitHub doesn't track changes on the target branch after the PR is opened. There are some suggested fixes here: https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch

Hm, I tried a few but I'm not getting too much success. I may close this PR and re-open it with new changes. I think it's because I cloned into my fork's `master` to update the files, when in reality I probably should've taken a different approach. Give me a moment to test things out!

**Comment by m-allanson** [comment]
> I may close this PR and re-open it with new changes. 

That's often the quickest way to fix these sorts of things 😅. 

> I think it's because I cloned into my fork's master

Feel free to clone `stylelint/stylelint` and create branches straight from this repo instead of your fork. Though I sometimes prefer working from a forked repo too. 

> Give me a moment to test things out!

Absolutely, no rush.

**Comment by malsf21** [comment]
Closed in favour of #4878, which is the same solution but with a cleaner git history.

**Pull request closed without merging**